### PR TITLE
feat(gateway-types/class_hash): add class hash calculation for Sierra classes

### LINKED ISSUE #903: Parse sierra classes

### PR DESCRIPTION
**Pull request by kkovaacs**
This PR adds class hash calculation code for Sierra classes.

Class hash calculation now first parses the JSON blob to decide if it's a Cairo or a Sierra class definition and then calls the appropriate class hash calculation function.

Sierra class hash calculation is significantly simpler than for Cairo classes, since it does _not_ involve serializing JSON and calculating hashes for the JSON output. Instead, ABI is handled as a string and all other relevant parts of the class definition are transformed into Felts and hashed using Poseidon.

For reference see [the documentation](https://docs.starknet.io/documentation/architecture_and_concepts/Contracts/class-hash/) and the [Cairo implementation](https://github.com/starkware-libs/cairo-lang/blob/12ca9e91bbdc8a423c63280949c7e34382792067/src/starkware/starknet/core/os/contract_class/contract_class.cairo#L42).

Closes #903 #904

### CODE REVIEW

**Comment by CHr15F0x** [comment]
LGTM, will re-review once not a draft anymore

File: crates/gateway-types/src/class_hash.rs [hunk 1]
```diff
@@ -231,6 +280,76 @@ fn compute_class_hash0(mut contract_definition: json::ContractDefinition<'_>) ->
...
+    if contract_definition.contract_class_version != "0.1.0" {
+        anyhow::bail!("Unsupported Sierra class version");
+    }
+
+    let mut hash = PoseidonHasher::default();
+
+    const SIERRA_VERSION: Felt = felt_bytes!(b"CONTRACT_CLASS_V0.1.0");
+    hash.write(SIERRA_VERSION.into());
+
+    // It is important process the different entrypoint hashchains in correct order.
```
Comment (CHr15F0x) [inline crates/gateway-types/src/class_hash.rs:308]: nit, line 174 (223) too

```suggestion
    // It is important to process the different entrypoint hashchains in correct order.
```

**Review by CHr15F0x — APPROVED** [review]
LGTM :+1:

File: crates/gateway-types/src/class_hash.rs (hunk 1 above)
Comment (kkovaacs) [reply crates/gateway-types/src/class_hash.rs]: Thanks, done.

**Pull request merged**